Provide and use signaling nans

bangerth opened the pull request:
This is an adaptation of @tjhei 's original patch to ASPECT (https://github.com/geodynamics/aspect/pull/441) that introduced the ability to initialize arrays to signaling NaNs. The idea is that this pre-sets uninitialized values to something that triggers a processor exception whenever this (uninitialized) data is accidentally used in any kind of arithmetic.

I generalized Timo's way of generating signaling NaNs a bit, doing away with the complete specializations of the function and instead allowing myself the luxury of partially specialized class templates. This is more extensible. 

The second commit uses this new facility to pre-set the elements of the various internal FEValues structures to these invalid values. This helped me find the issue addressed by #1908.

My test machine does not have sufficiently modern support for floating point exceptions, so they are switched off and I cannot see whether this patch actually triggers any of the exceptions this work hopes to capture. I suppose we will simply see what we find on other regression testers once this patch goes in. (Unless, of course, someone for whom `DEAL_II_HAVE_FP_EXCEPTIONS` is defined in `$build/include/deal.II/base/config.h` wants to run the testsuite with this patch.)

kronbichler commented:
This looks good. I have only one comment: Is `Utilities::signaling_nan()` the most appropriate namespace or should it rather be `numbers::signaling_nan()` where we have checks like `is_nan()`? I'd be happy to hear some opinions on this.

tjhei commented:
I remember that constructing these objects without triggering floating point exceptions was somewhat tricky, so we need to test this before merging.

tjhei commented:
Can you please add a test that constructs these objects? For SymmetricTensor I am getting compile errors
```
In file included from /ssd/deal-build/installed/include/deal.II/base/function.h:25:0,
                 from /ssd/deal-build/installed/include/deal.II/grid/grid_generator.h:24,
                 from /ssd/deal-build/installed/examples/step-3/step-3.cc:29:
/ssd/deal-build/installed/include/deal.II/base/symmetric_tensor.h: In instantiation of âconst unsigned int dealii::SymmetricTensor<1, 3>::n_independe
nt_componentsâ:
/ssd/deal-build/installed/include/deal.II/base/symmetric_tensor.h:990:1:   instantiated from âdealii::SymmetricTensor<1, 3>â
/ssd/deal-build/installed/examples/step-3/step-3.cc:636:3:   instantiated from here
/ssd/deal-build/installed/include/deal.II/base/symmetric_tensor.h:518:7: error: incomplete type âdealii::internal::SymmetricTensorAccessors::StorageT
ype<1, 3, double>â used in nested name specifier
/ssd/deal-build/installed/include/deal.II/base/symmetric_tensor.h: In instantiation of âdealii::SymmetricTensor<1, 3>â:
/ssd/deal-build/installed/examples/step-3/step-3.cc:636:3:   instantiated from here
/ssd/deal-build/installed/include/deal.II/base/symmetric_tensor.h:990:1: error: size of array is not an integral constant-expression

```

bangerth commented:
On 11/26/2015 10:25 AM, Martin Kronbichler wrote:
> This looks good. I have only one comment: Is |Utilities::signaling_nan()| the
> most appropriate namespace or should it rather be |numbers::signaling_nan()|
> where we have checks like |is_nan()|? I'd be happy to hear some opinions on this.

Oh yes, that's an excellent point I had meant to ask in my pull request but 
forgot. I think `numbers` actually makes more sense now that you mention it. 
What do others feel?

bangerth commented:
On 11/28/2015 12:34 PM, Timo Heister wrote:
> I remember that constructing these objects without triggering floating point
> exceptions was somewhat tricky, so we need to test this before merging.

It wasn't the creation of these objects, but instead switching on floating 
point exceptions (which we already do, and have a configure time test for).

tjhei commented:
> What do others feel?

Agreed.

bangerth commented:
Good spot, @tjhei . There was a bug in a copy constructor of SymmetricTensor that led to that compiler error. I pushed another patch for this, in addition to one that adds tests.

I've also moved everything into namespace `numbers`.

tjhei commented:
Thanks for the changes. A couple of problems:
1. You can not print signaling NaNs without triggering them. I would change the tests to just create the objects. (that seems to work)
2. ```Table<2,double>::reinit(m,n,signaling_nan)``` triggers a floating point exception, so you can not init ```shape_values``` that way.
3. I haven't checked the other places yet.

bangerth commented:
> 1. You can not print signaling NaNs without triggering them. I would change
> the tests to just create the objects. (that seems to work)

Aw, that's lame. But you're right. I've updated the tests.


> 2. |Table<2,double>::reinit(m,n,signaling_nan)| triggers a floating point
> exception, so you can not init |shape_values| that way.

I double checked this, and you're right, but the reason is that the function 
looks like this:
```
   void reinit (const unsigned int size1,
                const unsigned int size2,
                const bool         omit_default_initialization = false);
```
What fails is not the initialization, but the conversion from sNaN to bool. 
I've added another test that does the initialization via an array and works as 
expected.


> 3. I haven't checked the other places yet.
Let me know when you're through!

tjhei commented on tests/base/signaling_nan_table.cc:
```diff
@@ -0,0 +1,61 @@
+void check ()
+{
+  Table<2,double> t;
+  const double x = numbers::signaling_nan<double>();
```
you are not using ```T``` here, so ```float``` is not tested.

bangerth replied:
Good point. Fixed.

tjhei commented:
Can you take over https://github.com/tjhei/dealii/commit/26c9b419152bed5c517667d9b02ce6bda973240c ?

tjhei commented:
and the other commit at https://github.com/tjhei/dealii/commits/pr1909

Tests are passing...

kronbichler commented:
I'm fine with the patch. As soon as @tjhei's concerns are fixed, this can be merged.

bangerth commented:
Thanks all for testing. I had tested this on a machine that didn't have floating point exceptions switched on.

bangerth commented:
Let's see what the testers say about this...

tjhei commented on tests/base/signaling_nan_table.cc:
```diff
@@ -0,0 +1,61 @@
+  const T x = numbers::signaling_nan<double>();
+  const T array[6] = { x,x,x,x,x,x };
+
+  t.reinit (2,3, &x);
```
This looks weird to me. I would prefer you use ```t.reinit(2,3);t.fill(x);``` like we now do in FE

bangerth replied:
arg, I managed to mess it up :-( I meant of course to initialize with `array` and not `x`. Will fix immediately.

bangerth commented:
In reference to #1219.
